src/cpp/rtps/persistence/SQLite3PersistenceService.cpp

```diff
@@ -29,29 +29,138 @@ namespace eprosima {
+    {
+        return SQLITE_OK;
+    }
+
+    if ((from == 0 || from == 1) && to == 2)
+    {
+        return upgrade_version_1_to_version_2(db);
+    }
+
+    // unsupported upgrade path
+    logError(RTPS_PERSISTENCE, "Unsupported database upgrade from version " << from << " to version " << to);
+    return SQLITE_ERROR;
+}
+
+static sqlite3* open_or_create_database(const char* filename, bool update_schema)
 {
     sqlite3* db = NULL;
     int rc;
+    int version = 2;
 
     // Open database
-    int flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+    int flags = SQLITE_OPEN_READWRITE |
                 SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_SHAREDCACHE;
     rc = sqlite3_open_v2(filename, &db, flags, 0);
     if (rc != SQLITE_OK)
     {
-        sqlite3_close(db);
-        return NULL;
+        //probably the database does not exists. Create new and no need to upgrade schema
+        flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+                SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_SHAREDCACHE;
+        rc = sqlite3_open_v2(filename, &db, flags, 0);
+        if (rc != SQLITE_OK)
+        {
+            logError(RTPS_PERSISTENCE, "Unable to create persistence database " << filename);
+            sqlite3_close(db);
+            return NULL;
+        }
     }
-
+    else
+    {
+        // Find the database version and handle upgrades
+        int db_version = database_version(db);
+        if (db_version != version)
+        {
+            if (update_schema)
+            {
+                rc = upgrade(db, db_version, version);
+                if (rc != SQLITE_OK)
+                {
+                    sqlite3_close(db);
+                    return NULL;
+                }
+            }
+            else
+            {
+                logError(RTPS_PERSISTENCE, "Old schema version " << db_version << " on database " << filename
+                    << ". Set property dds.persistence.update_schema to force automatic schema upgrade");
+                return NULL;
+            }
+            
+        }
+    }
+    
     // Create tables if they don't exist
     const char* create_statement = R"(
-CREATE TABLE IF NOT EXISTS writers(
-    guid text,
-    seq_num integer,
-    instance binary(16),
-    payload blob,
-    PRIMARY KEY(guid, seq_num DESC)
+PRAGMA user_version = 2;
+PRAGMA foreign_keys = ON;
+
+CREATE TABLE IF NOT EXISTS writers_states(
+    guid TEXT PRIMARY KEY, -- CHECK(guid REGEXP '([0-9a-fA-F]{1,2}\.){11}[0-9a-fA-F]{1,2}\|([0-9a-fA-F]{1,2}\.){3}[0-9a-fA-F]{1,2}'),
+    last_seq_num INTEGER CHECK(last_seq_num > 0)
+) WITHOUT ROWID;
+
+CREATE TABLE IF NOT EXISTS writers_histories(
+    guid TEXT, -- CHECK(guid REGEXP '([0-9a-fA-F]{1,2}\.){11}[0-9a-fA-F]{1,2}\|([0-9a-fA-F]{1,2}\.){3}[0-9a-fA-F]{1,2}'),
+    seq_num INTEGER CHECK(seq_num > 0),
+    instance BLOB CHECK(length(instance)=16),
+    payload BLOB,
+    PRIMARY KEY(guid, seq_num DESC), 
+    FOREIGN KEY (guid)
+        REFERENCES writers_states(guid)
```

Comment:
A global `static const std::string`?
Maybe even splitting the statements to be reused between other globals.

Reply:
I agree with the global in this case, as the database will always be created with the latest schema, so there will ever be a single statement to maintain. The splitting, on the other hand, will probably make the maintenance harder on the long run. The complete statements are easier to understand IMHO.